src/common/format.ts

```diff
@@ -19,6 +19,14 @@ export const F_EXPORTED: 'e' = 'e';
 export const F_INFERRED: '.' = '.';
 export const F_OMITTED: ',' = ',';
 
+/* all allowed flags */
+export type F_FLAG = typeof F_READONLY | typeof F_ABSTRACT | typeof F_PUBLIC
+    | typeof F_PRIVATE | typeof F_PROTECTED | typeof F_PROPERTY | typeof F_METHOD
+    | typeof F_STATIC | typeof F_CLASS | typeof F_INTERFACE | typeof F_FUNCTION
+    | typeof F_ARROW_FUNCTION | typeof F_OPTIONAL | typeof F_REST | typeof F_ASYNC
+    | typeof F_EXPORTED | typeof F_INFERRED | typeof F_OMITTED | "";
+
+export type F_FLAGS = `${F_FLAG| ''}`;
```

Comment:
This is really cool, I'd love to use template string typing for enforcing flags! However, I don't think this type works. See this TS Playground example for why: https://www.typescriptlang.org/play?#code/MYewdgzgLgBAYgfQEoFECCARA8gOQDICaMAvDAOQBOZA3AFCiSyKqYDqSAkgCoonkDuNWlACeABwCm8BHDxoA4n1GSQAM2kts+IgB8YyiWo3oM7bijoHpshQGU+AAwAkAb0Q3FesmQC+DugA2ErCqAQCGAOYQAFzWcvL2pJSC1EA

Later on you use this type for the metadata structures which are strings that have _multiple_ flags in them: 

![image](https://user-images.githubusercontent.com/409140/180672820-fb4c6df2-ef90-4966-be73-3096b6464132.png)

...but the `F_FLAGS` definition here doesn't seem to allow that. Is there a way we can make this type accept multiple flags (and in any order) using string template types? I'd love it if so :-D

Reply:
we need to wait this unfortunately

https://github.com/microsoft/TypeScript/issues/41160

Reply:
"fixed" with this

```typescript
export type F_FLAGS = `${F_FLAG| ''}${F_FLAG| ''}${F_FLAG| ''}${string}`; // this type validate only the first 3 flags
```

this is the maximum that we can do with the current version of typescript. (validate the first 3, after accept any string)